src/NuGet.Services.AzureSearch/SearchService/AuxiliaryFileClient.cs

```diff
@@ -0,0 +1,149 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.WindowsAzure.Storage;
+using Newtonsoft.Json;
+using NuGet.Indexing;
+using NuGetGallery;
+
+namespace NuGet.Services.AzureSearch.SearchService
+{
+    public class AuxiliaryFileClient : IAuxiliaryFileClient
+    {
+        private readonly ICloudBlobClient _cloudBlobClient;
+        private readonly IOptionsSnapshot<SearchServiceConfiguration> _options;
+        private readonly ILogger<AuxiliaryFileClient> _logger;
+        private readonly Lazy<ICloudBlobContainer> _lazyContainer;
+
+        public AuxiliaryFileClient(
+            ICloudBlobClient cloudBlobClient,
+            IOptionsSnapshot<SearchServiceConfiguration> options,
+            ILogger<AuxiliaryFileClient> logger)
+        {
+            _cloudBlobClient = cloudBlobClient ?? throw new ArgumentNullException(nameof(cloudBlobClient));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            _lazyContainer = new Lazy<ICloudBlobContainer>(
+                () => _cloudBlobClient.GetContainerReference(_options.Value.AuxiliaryDataStorageContainer));
+        }
+
+        private ICloudBlobContainer Container => _lazyContainer.Value;
+
+        public async Task<AuxiliaryFileResult<Downloads>> LoadDownloadsAsync(string etag)
+        {
+            return await LoadAuxiliaryFileAsync(
+                _options.Value.AuxiliaryDataStorageDownloadsPath,
+                etag,
+                loader =>
+                {
+                    var downloads = new Downloads();
+                    downloads.Load(
+                        name: null,
+                        loader: loader,
+                        logger: _logger);
+                    return downloads;
+                });
+        }
+
+        public async Task<AuxiliaryFileResult<HashSet<string>>> LoadVerifiedPackagesAsync(string etag)
+        {
+            return await LoadAuxiliaryFileAsync(
+                _options.Value.AuxiliaryDataStorageVerifiedPackagesPath,
+                etag,
+                loader => VerifiedPackages.Load(
+                    fileName: null,
+                    loader: loader,
+                    logger: _logger));
+        }
+
+        private async Task<AuxiliaryFileResult<T>> LoadAuxiliaryFileAsync<T>(
+            string blobName,
+            string etag,
+            Func<ILoader, T> loadData) where T : class
+        {
+            _logger.LogInformation("Attempted to load blob {BlobName} with etag {ETag}.", blobName, etag);
+
+            var stopwatch = Stopwatch.StartNew();
+            var blob = Container.GetBlobReference(blobName);
```

Comment:
Can we use the `TelemetryClient` here? It'd be useful to record these durations as metrics too.

Reply:
Tracked https://github.com/NuGet/NuGetGallery/issues/6664.